File: frontend/src/api/services/bookingService.ts
import api from '../axios';
import type { BookingRequest, FieldCalendar } from '../../types';

export interface CreateBookingRequest {
    fieldId: number;
    teamId: number;
    date: string;
    startTime: string;
    endTime: string;
    notes?: string;
}

export interface CalendarQueryParams {
    startDate: string;
    endDate: string;
}

export interface BlockSlotRequest {
    fieldId: number;
    date: string;
    startTime: string;
    endTime: string;
    reason?: string;
}

// Booking service functions - Real API calls
export const bookingService = {
    /**
     * Create a new booking request
     */
    createBooking: async (data: CreateBookingRequest): Promise<BookingRequest> => {
        const response = await api.post<BookingRequest>('/bookings', data);
        return response.data;
    },

    /**
     * Get booking by ID
     */
    getBookingById: async (bookingId: number): Promise<BookingRequest | undefined> => {
        try {
            const response = await api.get<BookingRequest>(`/bookings/${bookingId}`);
            return response.data;
        } catch {
            return undefined;
        }
    },

    /**
     * Get pending bookings for a field
     */
    getPendingBookings: async (fieldId: number): Promise<BookingRequest[]> => {
        const response = await api.get<BookingRequest[]>(`/bookings/field/${fieldId}`);
        return response.data;
    },

    /**
     * Get pending bookings for all owner's fields
     */
    getOwnerPendingBookings: async (): Promise<BookingRequest[]> => {
        const response = await api.get<BookingRequest[]>('/bookings/owner/pending');
        return response.data;
    },

    /**
     * Approve booking request
     */
    approveBooking: async (bookingId: number): Promise<BookingRequest> => {
        const response = await api.put<BookingRequest>(`/bookings/${bookingId}/approve`);
        return response.data;
    },

    /**
     * Reject booking request
     */
    rejectBooking: async (bookingId: number, reason?: string): Promise<BookingRequest> => {
        const response = await api.put<BookingRequest>(`/bookings/${bookingId}/reject`, { reason });
        return response.data;
    },

    /**
     * Cancel booking
     */
    cancelBooking: async (bookingId: number): Promise<BookingRequest> => {
        const response = await api.put<BookingRequest>(`/bookings/${bookingId}/cancel`);
        return response.data;
    },

    /**
     * Get field calendar
     */
    getFieldCalendar: async (fieldId: number, params: CalendarQueryParams): Promise<FieldCalendar[]> => {
        const response = await api.get<FieldCalendar[]>(`/fields/${fieldId}/calendar`, { params });
        return response.data;
    },

    /**
     * Block a calendar slot
     */
    blockSlot: async (data: BlockSlotRequest): Promise<FieldCalendar> => {
        const response = await api.post<FieldCalendar>(`/fields/${data.fieldId}/calendar/block`, {
            date: data.date,
            startTime: data.startTime,
            endTime: data.endTime,
        });
        return response.data;
    },

    /**
     * Unblock a calendar slot
     */
    unblockSlot: async (calendarId: number): Promise<FieldCalendar> => {
        const response = await api.put<FieldCalendar>(`/calendar/${calendarId}/unblock`);
        return response.data;
    },

    /**
     * Get bookings for a team
     */
    getTeamBookings: async (teamId: number): Promise<BookingRequest[]> => {
        const response = await api.get<BookingRequest[]>(`/bookings/team/${teamId}`);
        return response.data;
    },
};

export default bookingService;

File: backend/app/services/booking_service.py
"""
BookingService - Booking request business logic.
Maps to BookingController in class diagram.
"""
from typing import List, Optional
from datetime import datetime, date, time
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.booking_repository import BookingRepository
from app.repositories.field_repository import CalendarRepository
from app.models.booking import BookingRequest
from app.models.field import FieldCalendar
from app.models.enums import BookingStatus, CalendarStatus


class BookingService:
    """Service handling booking business logic."""
    
    def __init__(self, db: AsyncSession):
        self.db = db
        self.booking_repo = BookingRepository(db)
        self.calendar_repo = CalendarRepository(db)
    
    async def create_booking(
        self,
        field_id: int,
        team_id: int,
        requester_id: int,
        booking_date: date,
        start_time: time,
        end_time: time,
        notes: str = None,
    ) -> BookingRequest:
        """Create a booking request."""
        booking = BookingRequest(
            field_id=field_id,
            team_id=team_id,
            requester_id=requester_id,
            date=booking_date,
            start_time=start_time,
            end_time=end_time,
            status=BookingStatus.PENDING,
            notes=notes,
        )
        await self.booking_repo.save(booking)
        await self.booking_repo.commit()
        return booking
    
    async def get_booking_by_id(self, booking_id: int) -> Optional[BookingRequest]:
        """Get booking by ID."""
        return await self.booking_repo.find_by_id(booking_id)
    
    async def get_bookings_by_field(
        self,
        field_id: int,
        status: BookingStatus = None
    ) -> List[BookingRequest]:
        """Get bookings for a field."""
        return await self.booking_repo.find_by_field(field_id, status)
    
    async def get_bookings_by_team(self, team_id: int) -> List[BookingRequest]:
        """Get bookings for a team."""
        return await self.booking_repo.find_by_team(team_id)
    
    async def approve_booking(self, booking: BookingRequest) -> bool:
        """Approve a booking request."""
        booking.status = BookingStatus.CONFIRMED
        booking.processed_at = datetime.utcnow()
        
        # Create calendar slot as booked
        calendar_slot = FieldCalendar(
            field_id=booking.field_id,
            date=booking.date,
            start_time=booking.start_time,
            end_time=booking.end_time,
            status=CalendarStatus.BOOKED,
            booking_id=booking.booking_id,
        )
        await self.calendar_repo.save(calendar_slot)
        
        await self.booking_repo.update(booking)
        await self.booking_repo.commit()
        return True
    
    async def reject_booking(self, booking: BookingRequest) -> bool:
        """Reject a booking request."""
        booking.status = BookingStatus.REJECTED
        booking.processed_at = datetime.utcnow()
        
        await self.booking_repo.update(booking)
        await self.booking_repo.commit()
        return True
    
    async def cancel_booking(self, booking: BookingRequest) -> bool:
        """Cancel a booking."""
        booking.status = BookingStatus.CANCELLED
        booking.processed_at = datetime.utcnow()
        
        await self.booking_repo.update(booking)
        await self.booking_repo.commit()
        return True
